projects/uriparser/uri_parse_fuzzer.cc

```diff
@@ -0,0 +1,119 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cstddef>
+#include <cstring>
+#include <string>
+#include <vector>
+#include <fuzzer/FuzzedDataProvider.h>
+
+using std::string;
+#include "uriparser/include/uriparser/Uri.h"
+#include "uriparser/include/uriparser/UriIp4.h"
+
+class UriParserA {
+ public:
+  UriParserA() { memset((void *)&uri_, 0, sizeof(uri_)); }
+  ~UriParserA() { uriFreeUriMembersA(&uri_); }
+
+  UriUriA *get_mutable_uri() { return &uri_; }
+  UriUriA *get_uri() const { return const_cast<UriUriA *>(&uri_); }
+
+ private:
+  UriUriA uri_;
+};
+
+void Escapes(const std::string &uri) {
+  const char *first = uri.c_str();
+  // A new line char takes 6 char to encode.
+  // Use a vector to make a C string.
+  std::vector<char> buf1(uri.size() * 6 + 1);
+  std::vector<char> buf2(uri.size() * 3 + 1);
+
+  char *result;
+  result = uriEscapeA(first, &buf1[0], URI_TRUE, URI_TRUE);
+  result = uriEscapeA(first, &buf1[0], URI_FALSE, URI_TRUE);
+  result = uriEscapeA(first, &buf2[0], URI_TRUE, URI_FALSE);
+  result = uriEscapeA(first, &buf2[0], URI_FALSE, URI_FALSE);
+  uriUnescapeInPlaceA(&buf1[0]);
+  uriUnescapeInPlaceA(&buf2[0]);
+}
+
+void FileNames(const std::string &uri) {
+  const size_t size = 8 + 3 * uri.size() + 1;
+  std::vector<char> buf(size);
+
+  uriUnixFilenameToUriStringA(uri.c_str(), &buf[0]);
+  uriWindowsFilenameToUriStringA(uri.c_str(), &buf[0]);
+  uriUriStringToUnixFilenameA(uri.c_str(), &buf[0]);
+  uriUriStringToWindowsFilenameA(uri.c_str(), &buf[0]);
+}
+
+int uriParseIpFourAddressA(unsigned char *octetOutput, const char *first,
+                           const char *afterLast);
+
+void Ipv4(const std::string &s) {
+  const char *cstr = s.c_str();
+  unsigned char result[4] = {};
+  uriParseIpFourAddressA(result, cstr, &cstr[s.size()]);
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
```

Comment:
This fuzzer does a lot.  I think it might make sense to split it into a number of much smaller fuzzers in another PR, do you?

Reply:
Sure, SGTM. I can get started on that as soon as this is merged.